servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/NewRoundRobinLoadBalancer.java

```diff
@@ -444,6 +441,35 @@ public Single<C> newConnection(@Nullable final ContextMap context) {
         return defer(() -> selectConnection0(c -> true, context, true).shareContextOnSubscribe());
     }
 
+    private Single<C> selectConnection0(final Predicate<C> selector, @Nullable final ContextMap context,
+                                        final boolean forceNewConnectionAndReserve) {
+        final List<Host<ResolvedAddress, C>> currentHosts = this.usedHosts;
+        // It's possible that we're racing with updates from the `onNext` method but since it's intrinsically
+        // racy it's fine to do these 'are there any hosts at all' checks here using the total host set.
+        if (currentHosts.isEmpty()) {
+            return isClosedList(currentHosts) ? failedLBClosed(targetResource) :
+                    // This is the case when SD has emitted some items but none of the hosts are available.
+                    failed(Exceptions.StacklessNoAvailableHostException.newInstance(
+                            "No hosts are available to connect for " + targetResource + ".",
+                            NewRoundRobinLoadBalancer.class, "selectConnection0(...)"));
+        }
+
+        Single<C> result = hostSelector.selectConnection(selector, context, forceNewConnectionAndReserve);
+        if (healthCheckConfig != null) {
+            result = result.beforeOnError(exn -> {
+                if (exn instanceof Exceptions.StacklessNoActiveHostException && allUnhealthy(currentHosts)) {
+                    final long currNextResubscribeTime = nextResubscribeTime;
+                    if (currNextResubscribeTime >= 0 &&
+                            healthCheckConfig.executor.currentTime(NANOSECONDS) >= currNextResubscribeTime &&
+                            nextResubscribeTimeUpdater.compareAndSet(this, currNextResubscribeTime, RESUBSCRIBING)) {
+                            subscribeToEvents(true);
+                    }
+                }
+            });
+        }
```

Comment:
This part has changed: we no longer have access to the knowledge if whether a host was even selected other than the type of exception that is returned. We could do some stuff like return `null` instead of a `Single.failed(..)` in the case that no host was picked but that API is pretty icky.

Reply:
Agreed, API doesn't favor `null`. For internal interface it's ok to use `null` as internal hack, but if we make it public in the future, it will be tricky for users. I'm ok with intercepting an exception here bcz this is a rare case when client turns into that state.